### ISSUE TITLE
RuntimeException: Cannot find an entity at (...) PackagingElement.getThisEntity

### ISSUE THREAD

**Issue by kriegaex** [JaneWardSandy/eclipse-pde-partial-idea#106]
Since 1.5.2, I often get this error in IDEA, even if no Eclipse PDE project is open at all. I just opened a totally unrelated Java & Groovy project.

```text
java.lang.RuntimeException: Cannot find an entity
	at com.intellij.packaging.elements.PackagingElement.getThisEntity(PackagingElement.java:155)
	at com.intellij.packaging.impl.elements.FileOrDirectoryCopyPackagingElement.getMyFilePath(FileOrDirectoryCopyPackagingElement.java:120)
	at com.intellij.packaging.impl.elements.FileOrDirectoryCopyPackagingElement.isEqualTo(FileOrDirectoryCopyPackagingElement.java:58)
	at com.intellij.packaging.impl.elements.FileCopyPackagingElement.isEqualTo(FileCopyPackagingElement.java:72)
	at com.intellij.packaging.elements.CompositePackagingElement.myAddOrFindChild(CompositePackagingElement.java:68)
	at com.intellij.packaging.elements.CompositePackagingElement.lambda$addOrFindChild$0(CompositePackagingElement.java:29)
	at com.intellij.packaging.elements.PackagingElement.update(PackagingElement.java:137)
	at com.intellij.packaging.elements.CompositePackagingElement.addOrFindChild(CompositePackagingElement.java:28)
	at cn.varsa.idea.pde.partial.plugin.helper.ModuleHelper.setCompileArtifact(ModuleHelper.kt:126)
	at cn.varsa.idea.pde.partial.plugin.helper.ModuleHelper.setCompileArtifact(ModuleHelper.kt:77)
	at cn.varsa.idea.pde.partial.plugin.helper.ModuleHelper.setCompileArtifact$default(ModuleHelper.kt:66)
	at cn.varsa.idea.pde.partial.plugin.helper.ModuleHelper.resetCompileArtifact(ModuleHelper.kt:63)
	at cn.varsa.idea.pde.partial.plugin.config.BundleManagementService$onFinished$1$1.resolve(BundleManagementService.kt:133)
	at cn.varsa.idea.pde.partial.plugin.support.BackgroundResolvable$backgroundResolve$6.run(BackgroundResolvable.kt:25)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:429)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$6(CoreProgressManager.java:480)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by JaneWardSandy** [comment]
Can you provide a sample project that reproduces the problem?
Same issue #102

**Comment by kriegaex** [comment]
I do not have one which always reproduces it when opening it multiple times. But chances are, that https://github.com/joker1/m2eclipse-ajdt might cause the problem during an initial Maven import or after not having opened the project for a while, maybe causing some time of re-import or re-indexing. I really have no idea which actions or circumstances really cause the problem. I saw the error once today, but could not find a way to reproduce it.

**Comment by JaneWardSandy** [comment]
🤔

As you said, I tried to open m2eclipse-ajdt and pde-related projects to reproduce this problem, but I couldn't reproduce it once. Whether it was opening the two projects of Mavan and PDE at the same time, Gradle or IntellIj + PDE, also Or add Facet of PDE to Maven tycho.

For example, #102, I couldn't reproduce a similar error.

I would appreciate it if you could try to fix this problem and submit a PR.

**Comment by kriegaex** [comment]
Does the stack trace help you to guess what the problem might be in `ModuleHelper`? My feeling when getting this error that everytime I got it, I was doing something unrelated. Maybe the key is somewhere in `BackgroundResolvable`, which is also in the stack trace. Maybe there is a race condition or concurrency issue. WDYT?

> I would appreciate it if you could try to fix this problem and submit a PR.

Sorry, I don't do IDEA plugins and do not speak Kotlin either. I am just a user.

**Comment by JaneWardSandy** [comment]
Because the problem has not been reproduced, and there is no need to use Artefact to build jar after using Maven, a Workaround is provided here to skip this error.

In the `Add Framework Support` and `Facet` settings, you can now choose not to update (skip) the setting build path and artifacts.

<img width="585" alt="未命名" src="https://user-images.githubusercontent.com/15345355/225516793-b297914c-76f3-4118-8829-6c5caad778dd.png">

<img width="602" alt="截屏2023-03-16 12 36 40" src="https://user-images.githubusercontent.com/15345355/225516798-42daf962-5eea-41b7-a269-309b4aeb25d7.png">

**Comment by kriegaex** [comment]
Sorry, this does not seem to have helped at all in my case. I am on version 1.6.0.1. I just re-opened project https://github.com/eclipse/org.aspectj, which does not have any facets defined at all throughout the many modules, neither OSGi nor PDE Partial ones. Before, no other project was open, I opened the single project from the recent projects list which appears, when you close the last open project.

There is, however, a pop-up asking me whether I like to configure OSGi,  because some OSGi-related stuff seems to have been detected in the project. I did not, like I said, add any OSGi facets to the project. Maybe this is a way for you to reproduce the problem. Maybe you were looking for something in an OSGi project with already existing OSGi facets, not in one without such facets.

**Comment by JaneWardSandy** [comment]
I follow the steps below, but still can't reproduce:
1. Open `org.aspectj` project
2. `mvn package`
3. Close the project after completing `package`
4. In the recent project dialog that pops up from IDEA, click the `org.aspectj` project again
5. Wait for further pop-up prompts

At this step, it is the same as you, there is a prompt to detect `OSGi framework`, but IDEA does not report an error
<img width="627" alt="image" src="https://user-images.githubusercontent.com/15345355/226182289-ed4918e5-8935-45c2-9e4b-0397c12fbbe9.png">
<img width="937" alt="image" src="https://user-images.githubusercontent.com/15345355/226182389-556c3556-9cca-4c23-9793-bd1e66f83235.png">

The `OSGi framework` instrumentation is an instrumentation provided by the [`OSGi(Osmorc)` plugin](https://plugins.jetbrains.com/plugin/1816-osgi) from `Jetbrains`.
The detection provided by the current plug-in is `Eclipse PDE Partial framework`, modifying `Artifacts` will only be executed if the `Eclipse PDE Partial` Facet is added to the `Module`

I tried doing different things at step 4 above like
- Open `org.aspectj`
- Open the project without any Facets
- Open an existing `Eclipse PDE Partial` project
- Open an existing `OSGi` project
- Open a project that already has both `Eclipse PDE Partial` and `OSGi`
- Open the project that already has `Eclipse PDE Partial`, and add `OSGi` when detecting the suggestion, close the project after waiting and open it here
- Open an existing `OSGi` project and add `Eclipse PDE Partial` when detecting suggestions, wait and close the project and open it here

All operations failed to reproduce the error in your screenshot

If your project does not add the `Eclipse PDE Partial` Facet, the first line in the implementation of the `ModuleHelper#resetCompileArtifact` method will leave the method
If your project added the `Eclipse PDE Partial` Facet, but skipped updating `Artifacts`, it should exit on the second line of the method
https://github.com/JaneWardSandy/eclipse-pde-partial-idea/blob/159195a4b93964b333fb3508956fc1f66b1a3981/eclipse-pde-partial-idea/src/main/kotlin/cn/varsa/idea/pde/partial/plugin/helper/ModuleHelper.kt#L76-L82

For now, I can't reproduce your error. Maybe you can provide more complete reproduction steps, including what operations are performed in each step, sample projects, Facet in the project, etc.
Videos can also be provided if the description is not good (if it does not involve sensitive content)

**Comment by kriegaex** [comment]
The stack trace proves that I get the error, reproducible for you or not. And I can only reiterate that the error occurs, even though my project neither has any PDE nor OSGi facets. I do not need those two either in my project. Still, the plugin annoys me with its stack traces. There must be a way to fix that. I only installed the plugin for a different project, not for the one I mentioned above.

**Comment by JaneWardSandy** [comment]
I can understand your expectations. At present, I have released version 1.6.1. You can try the new version. As far as I reproduce the problem, it should have been fixed.

---

Here are the steps I have reproduced so far:

- In an Eclipse PDE Partial project, Facet has been added, and the Artifacts structure has been added by the plug-in.
- Remove all child nodes except the root node in `Artifacts`
- When clicking `OK`, IDEA will report an error. The error plug-in is `Java`, and the content is also to throw `Cannot find an entity` in `PackagingElement.getThisEntity`

<img width="1288" alt="截屏2023-03-20 11 29 23" src="https://user-images.githubusercontent.com/15345355/226251237-4d6b17d7-1fbe-484a-864b-73052551e8de.png">

- Close the project and open the project again in the recent list
- `Artifacts` should be rebuilt in `PostAvtivity`. At this time, an error will be reported, and the content of the error will be consistent with your screenshot.
- If there is no reconstruction of `Artifacts`, close the project and open it again. Right-click `Re-resolve workspace` in the `Project` view, and an error will still be reported in rebuilding `Artifacts`.
- Only after completely deleting `Artifacts` will there be no error when rebuilding.

---

The current code modification scheme is that the structure is no longer reconstructed on the original `Artifacts`. When the structure changes, the code directly removes `Artifacts` and creates a new one.

At present, there is no error in the test of this scheme, and the problem should be repaired.

**Comment by kriegaex** [comment]
Sorry for the belated feedback. The reason is threefold: lots of work, some health issues and me wanting to watch the situation with v1.6.1 for a while before regarding the issues as fixed. Since the update, I did not see that stack trace anymore, i.e. the issue seems to be fixed. Thank you so much for your feedback and support and for maintaining this plugin.
